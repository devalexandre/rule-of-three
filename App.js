import React from "react";
import { SafeAreaView, View, Platform, StatusBar } from "react-native";
import App from "./src/App";
import brandColor from "./src/colors";

export default () => (
  <SafeAreaView style={{ backgroundColor: brandColor, height: "100%" }}>
    <StatusBar barStyle="light-content" />
    {Platform.OS === "android" && (
      <View style={{ height: StatusBar.currentHeight }} />
    )}
    {Platform.OS === "ios" &&
      Platform.Version < 11 && (
        <View style={{ backgroundColor: brandColor, height: 20 }} />
      )}
    <App />
  </SafeAreaView>
);
