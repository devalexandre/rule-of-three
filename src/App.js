import React from "react";
import { StyleSheet, View, TextInput } from "react-native";
import calculate from "./calculate";

export default class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      a: 0,
      b: 0,
      c: 0,
      result: 0
    };
  }

  handleFormChange() {
    this.setState(({ a, b, c }) => ({
      a,
      b,
      c,
      result: calculate({ a, b, c })
    }));
  }

  handleChangeText(param, value) {
    const newState = { ...this.state };
    newState[param] = value;
    this.setState(newState);
    this.handleFormChange();
  }

  render() {
    return (
      <View style={styles.container}>
        <TextInput
          onChangeText={this.handleChangeText.bind(this, "a")}
          style={styles.input}
        />
        <TextInput
          onChangeText={this.handleChangeText.bind(this, "b")}
          style={styles.input}
        />
        <TextInput
          onChangeText={this.handleChangeText.bind(this, "c")}
          style={styles.input}
        />
        <TextInput
          style={styles.input}
          editable={false}
          value={this.state.result || ""}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    alignItems: "center",
    justifyContent: "center"
  },
  input: {
    borderColor: "#999",
    borderWidth: 1
  }
});
