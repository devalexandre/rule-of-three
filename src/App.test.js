import React from "react";
import App from "./App";

import { create } from "react-test-renderer";

const app = create(<App />);

it("renders without crashing", () => expect(app.toJSON()).toMatchSnapshot());

it("behaves as expected", () => {
  const expected = {
    a: 125,
    b: 100,
    c: 100,
    result: 80
  };
  const instance = app.getInstance();

  // TODO test event handlers instead?
  instance.handleChangeText("a", 125);
  instance.handleChangeText("b", 100);
  instance.handleChangeText("c", 100);

  expect(instance.state).toMatchObject(expected);
});
