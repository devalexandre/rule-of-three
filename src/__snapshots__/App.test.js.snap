// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "flex-start",
      "padding": "5%",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "margin": "auto",
        "maxWidth": 600,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#222",
          "fontSize": 32,
          "fontWeight": "bold",
        }
      }
    >
      Rule of Three
    </Text>
    <View>
      <View
        style={
          Object {
            "flex": -1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 16,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          onChangeText={[Function]}
          placeholder="Value 1"
          selectionColor="#7746e8"
          style={
            Array [
              Object {
                "borderColor": "#dedede",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#333",
                "fontSize": 24,
                "height": 40,
                "paddingHorizontal": 3,
                "paddingVertical": 8,
                "textAlign": "center",
                "width": "48%",
              },
              undefined,
            ]
          }
          underlineColorAndroid="#7746e8"
        />
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          onChangeText={[Function]}
          placeholder="Value 2 (e.g. 100%)"
          selectionColor="#7746e8"
          style={
            Array [
              Object {
                "borderColor": "#dedede",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#333",
                "fontSize": 24,
                "height": 40,
                "paddingHorizontal": 3,
                "paddingVertical": 8,
                "textAlign": "center",
                "width": "48%",
              },
              undefined,
            ]
          }
          underlineColorAndroid="#7746e8"
        />
      </View>
      <View
        style={
          Object {
            "flex": -1,
            "flexDirection": "row",
            "justifyContent": "space-between",
            "marginTop": 16,
          }
        }
      >
        <TextInput
          allowFontScaling={true}
          keyboardType="numeric"
          onChangeText={[Function]}
          placeholder="Value 3"
          selectionColor="#7746e8"
          style={
            Array [
              Object {
                "borderColor": "#dedede",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#333",
                "fontSize": 24,
                "height": 40,
                "paddingHorizontal": 3,
                "paddingVertical": 8,
                "textAlign": "center",
                "width": "48%",
              },
              undefined,
            ]
          }
          underlineColorAndroid="#7746e8"
        />
        <TextInput
          allowFontScaling={true}
          disabled={true}
          onFocus={[Function]}
          placeholder="Result"
          selectTextOnFocus={true}
          selectionColor="#7746e8"
          style={
            Array [
              Object {
                "borderColor": "#dedede",
                "borderRadius": 4,
                "borderWidth": 1,
                "color": "#333",
                "fontSize": 24,
                "height": 40,
                "paddingHorizontal": 3,
                "paddingVertical": 8,
                "textAlign": "center",
                "width": "48%",
              },
              Object {
                "fontWeight": "bold",
              },
            ]
          }
          underlineColorAndroid="#7746e8"
          value=""
        />
      </View>
    </View>
  </View>
</View>
`;
